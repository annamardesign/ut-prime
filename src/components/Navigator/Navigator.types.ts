import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';

export interface Props {
    keyField: string;
    field: string;
    title: string;
    resultSet: string;
    fetch: (params: {}) => Promise<{}>;
    className?: string;
    parentField?: string;
}

const styles = createStyles({
    component: {
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
