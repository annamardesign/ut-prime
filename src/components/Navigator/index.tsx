import React from 'react';
import clsx from 'clsx';
import DataSource from 'devextreme/data/data_source';
import CustomStore from 'devextreme/data/custom_store';
import TreeList, {Column} from 'devextreme-react/tree-list';

import {
    Styled,
    StyledType
} from './Navigator.types';

const Navigator: StyledType = ({
    classes,
    className,
    fetch,
    field,
    title,
    keyField,
    parentField = 'parents',
    resultSet
}) => {
    const tree = React.useMemo(() => new DataSource({
        store: new CustomStore({
            key: keyField,
            load: async() => (await fetch({}))[resultSet]
        })
    }), [fetch]);
    return (
        <TreeList
            height='100%'
            className={clsx(classes.component, className)}
            id="employees"
            dataSource={tree}
            focusedRowEnabled={true}
            defaultExpandedRowKeys={[null]}
            showRowLines={false}
            showBorders={false}
            columnAutoWidth={true}
            rootValue={null}
            keyExpr={keyField}
            parentIdExpr={parentField}
        >
            <Column dataField={field} caption={title} />
        </TreeList>
    );
};

export default Styled(Navigator);
