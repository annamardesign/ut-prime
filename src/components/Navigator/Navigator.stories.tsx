import React from 'react';
import { withReadme } from 'storybook-readme';

import Wrap from '../test/wrap';

// @ts-ignore: md file and not a module
import README from './README.md';
import Navigator from './index';

export default {
    title: 'Navigator',
    component: Navigator,
    decorators: [withReadme(README)]
};

const state = {
};

export const Basic: React.FC<{}> = () => <Wrap state={state}>
    <div style={{height: 500}}>
        <Navigator
            fetch={() => Promise.resolve({
                items: [...Array(50).keys()].map(number => ({
                    id: number,
                    name: `Item ${number}`,
                    parents: number > 10 ? number % 10 : undefined
                }))
            })}
            keyField='id'
            resultSet='items'
            field='name'
            title='Name'
        >
            <div>Navigation component</div>
        </Navigator>
    </div>
</Wrap>;
