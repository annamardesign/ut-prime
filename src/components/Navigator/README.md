# Navigator

## How to use

```jsx
import Navigator from 'ut-front-devextreme/core/Navigator';

<Navigator className='name' />;
```

## Props

- **className** - class applied to root `Navigator` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| className | string   |              | no         |
