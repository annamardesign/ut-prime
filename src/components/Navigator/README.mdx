# Navigator
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

Shows a hierarchical data as a tree, to enable
navigating collections of items related to that
tree.

Navigator is named after the `Navigation pane`
in Windows File Explorer.

## How to use

```jsx
import Navigator from 'ut-prime/core/Navigator';

<Navigator
    fetch={fetch}
    keyField='id'
    resultSet='items'
    field='name'
    title='Name'
>
    body
</Navigator>
```

<Canvas>
    <Story id="navigator--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
