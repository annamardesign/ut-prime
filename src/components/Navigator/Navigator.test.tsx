import React from 'react';
import {waitFor} from '@testing-library/react';

import { render } from '../test';
import Navigator from './index';

describe('<Navigator />', () => {
    it('render equals snapshot', async() => {
        const { getByTestId } = render(<Navigator
            fetch={() => Promise.resolve({
                items: [...Array(50).keys()].map(number => ({
                    id: number,
                    name: `Item ${number}`,
                    parents: number > 10 ? number % 10 : undefined
                }))
            })}
            keyField='id'
            resultSet='items'
            field='name'
            title='Name'
        />);
        await waitFor(() => expect(getByTestId('ut-front-test').querySelector('div.dx-treelist-text-content')).toBeTruthy());
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
