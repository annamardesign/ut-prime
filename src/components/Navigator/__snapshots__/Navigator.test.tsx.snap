// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigator /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="Navigator-component-1 dx-widget dx-visibility-change-handler dx-treelist"
    id="employees"
    role="presentation"
    style="height: 100%;"
  >
    <div
      aria-label="Tree list"
      class="dx-treelist-container dx-gridbase-container"
      role="treegrid"
    >
      <div
        class="dx-hidden"
      />
      <div
        class="dx-hidden"
      />
      <div
        class="dx-hidden"
      />
      <div
        class="dx-hidden"
      />
      <div
        class="dx-treelist-headers dx-treelist-nowrap"
        role="presentation"
      >
        <div
          class="dx-treelist-content dx-treelist-scroll-container"
          role="presentation"
        >
          <table
            class="dx-treelist-table dx-treelist-table-fixed"
            role="presentation"
          >
            <colgroup>
              <col />
            </colgroup>
            <tbody
              role="presentation"
            />
          </table>
        </div>
      </div>
      <div
        class="dx-treelist-rowsview dx-treelist-nowrap dx-empty"
        role="presentation"
        tabindex="0"
      >
        <div
          class="dx-treelist-content"
        >
          <table
            class="dx-treelist-table dx-treelist-table-fixed"
            role="presentation"
          >
            <colgroup>
              <col />
            </colgroup>
            <tbody
              role="presentation"
            >
              <tr
                class="dx-row dx-freespace-row dx-column-lines"
                role="presentation"
              >
                <td
                  style="text-align: left;"
                />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="dx-overlay dx-widget dx-state-invisible dx-visibility-change-handler dx-loadpanel"
      >
        <div
          aria-hidden="true"
          class="dx-overlay-content"
          style="width: 200px; height: 90px;"
        />
      </div>
      <div />
      <div
        class="dx-hidden"
      />
      <div
        class="dx-hidden"
      />
      <div
        class="dx-treelist-drag-header dx-treelist-text-content dx-widget dx-state-invisible"
      />
      <div
        class="dx-context-menu dx-has-context-menu dx-widget dx-visibility-change-handler dx-collection dx-treelist-container"
        role="presentation"
      />
      <div
        class="dx-header-filter-menu"
      />
      <div />
    </div>
  </div>
</div>
`;
